Project-Aurora/Project-Aurora/Devices/UnifiedHID/UnifiedHID.cs

```diff
@@ -500,4 +562,460 @@ public bool SetBottomLed(byte r, byte g, byte b)
+
+                    ctrl_device.OpenDevice();
+                    ctrl_device_leds.OpenDevice();
+
+                    bool success =
+                        rv_get_ctrl_report(0x0f) &&
+                        rv_set_ctrl_report(0x15) &&
+                        rv_wait_for_ctrl_device() &&
+                        rv_set_ctrl_report(0x05) &&
+                        rv_wait_for_ctrl_device() &&
+                        rv_set_ctrl_report(0x07) &&
+                        rv_wait_for_ctrl_device() &&
+                        rv_set_ctrl_report(0x0a) &&
+                        rv_wait_for_ctrl_device() &&
+                        rv_set_ctrl_report(0x0b) &&
+                        rv_wait_for_ctrl_device() &&
+                        rv_set_ctrl_report(0x06) &&
+                        rv_wait_for_ctrl_device() &&
+                        rv_set_ctrl_report(0x09) &&
+                        rv_wait_for_ctrl_device() &&
+                        rv_set_ctrl_report(0x0d) &&
+                        rv_wait_for_ctrl_device() &&
+                        rv_set_ctrl_report(0x13) &&
+                        rv_wait_for_ctrl_device();
+
+                    return (IsConnected = true);
+                }
+            }
+            catch (Exception exc)
+            {
+                Global.logger.LogLine($"Error when attempting to open UnifiedHID device:\n{exc}", Logging_Level.Error);
+            }
+
+            return false;
+        }
+
+        // We need to override Disconnect() too cause we have two hid devices open for this keyboard
+        public override bool Disconnect()
+        {
+            try
+            {
+                ctrl_device.CloseDevice();
+                ctrl_device_leds.CloseDevice();
+                return true;
+            }
+            catch (Exception exc)
+            {
+                Global.logger.LogLine($"Error when attempting to close UnifiedHID device:\n{exc}", Logging_Level.Error);
+            }
+            return false;
+        }
+
+        public override bool SetMultipleLEDColour(Dictionary<DeviceKeys, Color> keyColors)
+        {
+            try
+            {
+                if (!this.IsConnected)
+                    return false;
+
+                rv_rgb_map rgbMap = new rv_rgb_map();
+
+                foreach (KeyValuePair<DeviceKeys, Color> key in keyColors)
+                {
+
+
+                    Color clr = Color.FromArgb(255, Utils.ColorUtils.MultiplyColorByScalar(key.Value, key.Value.A / 255.0D));
+
+                    int index = DeviceKeyToRoccatVulcanIndex(key.Key);
+
+                    if (index == -1)
+                    {
+                        continue;
+                    }
+
+                    rgbMap.keys[index].r = clr.R;
+                    rgbMap.keys[index].g = clr.G;
+                    rgbMap.keys[index].b = clr.B;
+
+                }
+
+                rv_send_led_map(rgbMap);
```

Comment:
Wouldn't it be worth it to merge the rv_send_led_map into the keys loop so that you're not having to do two loops? You can do all the processing on a particular key in one go

Reply:
It'd also save having to allocate the intermediary rgbMap struct as you don't need to store that at all